keps/sig-storage/177-volume-snapshot/README.md

```diff
@@ -231,24 +251,288 @@ Here are the changes since the original design proposal:
+https://git.k8s.io/enhancements/keps/sig-architecture/20190731-production-readiness-review-process.md.
+
+The production readiness review questionnaire must be completed for features in
+v1.19 or later, but is non-blocking at this time. That is, approval is not
+required in order to be in the release.
+
+In some cases, the questions below should also have answers in `kep.yaml`. This
+is to enable automation to verify the presence of the review, and to reduce review
+burden and latency.
+
+The KEP must have a approver from the
+[`prod-readiness-approvers`](http://git.k8s.io/enhancements/OWNERS_ALIASES)
+team. Please reach out on the
+[#prod-readiness](https://kubernetes.slack.com/archives/CPNHUMN74) channel if
+you need any help or guidance.
+
+-->
+
+### Feature Enablement and Rollback
+
+_This section must be completed when targeting alpha to a release._
+
+* **How can this feature be enabled / disabled in a live cluster?**
+  - [x] Feature gate (also fill in values in `kep.yaml`)
+    - Feature gate name: VolumeSnapshotDataSource
+    - Components depending on the feature gate: kube-apiserver
+  - [ ] Other
+    - Describe the mechanism:
+    - Will enabling / disabling the feature require downtime of the control
+      plane?
+    - Will enabling / disabling the feature require downtime or reprovisioning
+      of a node? (Do not assume `Dynamic Kubelet Config` feature is enabled).
+
+* **Does enabling the feature change any default behavior?**
+  Yes. Enabling the feature gate will enable the volume snapshot support and disabling the feature gate will disable the volume snapshot support.
+
+* **Can the feature be disabled once it has been enabled (i.e. can we roll back
+  the enablement)?**
+  Yes. If the feature is disabled after it has been enabled, existing snapshots will still be there but they cannot be used as data source to provision new PVCs.
+
+* **What happens if we reenable the feature if it was previously rolled back?**
+  If we reenable the feature that was previously rolled back, the existing snapshots can be used again as data source to provision new PVCs.
+
+* **Are there any tests for feature enablement/disablement?**
+  The e2e framework does not currently support enabling or disabling feature
+  gates. However, unit tests in each component dealing with managing data, created
+  with and without the feature, are necessary. At the very least, think about
+  conversion tests if API types are being modified.
+
+### Rollout, Upgrade and Rollback Planning
+
+_This section must be completed when targeting beta graduation to a release._
+
+* **How can a rollout fail? Can it impact already running workloads?**
+  Try to be as paranoid as possible - e.g., what if some components will restart
+   mid-rollout?
+  The feature gate only affects kube-apiserver. If enabling the feature gate fails, already running workloads will not be able to use the snapshot feature.
+
+* **What specific metrics should inform a rollback?**
```

Comment:
Are there metrics that follow how many snapshots haven't been taken or how many pods using a snapshot datasource are stuck in pending or failing?  Both of those metrics seem pretty useful for this.

Reply:
Snapshots can't be mounted by a pod.  Snapshot can only be used as a data source to create a new volume.

There are metrics on failed create/delete snapshot operations.

Reply:
There are csi sidecar metrics on provisioner that show success/error/time for create volume operations. It currently doesn't distinguish between volumes created from a snapshot vs blank volume though.

Reply:
@deads2k there are two pieces of metrics in your question, I will try to address both of them:
1. how many snapshots haven't been taken?
This is a metric we think should be issued by snapshot controllers. We are current actively define the metrics to issue at [here](https://docs.google.com/document/d/1eW_5N7DTWPueLFQZkOnYxMiQ0Zd4pbze8AdkuO5XoCU/edit#heading=h.ygvrvicls5xh). One of them is to answer this question (# failures, # finished, and latencies)

2. how many pods using a snapshot datasource are stuck in pending or failing?
this is *not* part of snapshot controller's logic but rather sits in provisioner controller, and meanwhile, there are metrics around PVC provisioning as well.

Reply:
Updated KEP with this info.